x/staking/keeper/msg_server.go

```diff
@@ -372,3 +455,553 @@ func (k msgServer) Undelegate(goCtx context.Context, msg *types.MsgUndelegate) (
+	}
+
+	if validator.IsJailed() {
+		return nil, types.ErrValidatorJailed
+	}
+
+	ubd, found := k.GetUnbondingDelegation(ctx, delegatorAddress, valAddr)
+	if !found {
+		return nil, status.Errorf(
+			codes.NotFound,
+			"unbonding delegation with delegator %s not found for validator %s",
+			msg.DelegatorAddress, msg.ValidatorAddress,
+		)
+	}
+
+	// if this undelegation was from a liquid staking provider, the global and validator
+	// liquid counts should be incremented
+	tokens := msg.Amount.Amount
+	shares, err := validator.SharesFromTokens(tokens)
+	if err != nil {
+		return nil, err
+	}
+	if k.AccountIsLiquidStakingProvider(delegatorAddress) {
+		if err := k.SafelyIncreaseTotalLiquidStakedTokens(ctx, tokens, false); err != nil {
+			return nil, err
+		}
+		if err := k.SafelyIncreaseValidatorTotalLiquidShares(ctx, validator, shares); err != nil {
+			return nil, err
+		}
+	}
+
+	var (
+		unbondEntry      types.UnbondingDelegationEntry
+		unbondEntryIndex int64 = -1
+	)
+
+	for i, entry := range ubd.Entries {
+		if entry.CreationHeight == msg.CreationHeight {
+			unbondEntry = entry
+			unbondEntryIndex = int64(i)
+			break
+		}
+	}
+	if unbondEntryIndex == -1 {
+		return nil, sdkerrors.ErrNotFound.Wrapf("unbonding delegation entry is not found at block height %d", msg.CreationHeight)
+	}
+
+	if unbondEntry.Balance.LT(msg.Amount.Amount) {
+		return nil, sdkerrors.ErrInvalidRequest.Wrap("amount is greater than the unbonding delegation entry balance")
+	}
+
+	if unbondEntry.CompletionTime.Before(ctx.BlockTime()) {
+		return nil, sdkerrors.ErrInvalidRequest.Wrap("unbonding delegation is already processed")
+	}
+
+	// delegate back the unbonding delegation amount to the validator
+	_, err = k.Keeper.Delegate(ctx, delegatorAddress, msg.Amount.Amount, types.Unbonding, validator, false)
+	if err != nil {
+		return nil, err
+	}
+
+	amount := unbondEntry.Balance.Sub(msg.Amount.Amount)
+	if amount.IsZero() {
+		ubd.RemoveEntry(unbondEntryIndex)
+	} else {
+		// update the unbondingDelegationEntryBalance and InitialBalance for ubd entry
+		unbondEntry.Balance = amount
+		unbondEntry.InitialBalance = unbondEntry.InitialBalance.Sub(msg.Amount.Amount)
+		ubd.Entries[unbondEntryIndex] = unbondEntry
+	}
+
+	// set the unbonding delegation or remove it if there are no more entries
+	if len(ubd.Entries) == 0 {
+		k.RemoveUnbondingDelegation(ctx, ubd)
+	} else {
+		k.SetUnbondingDelegation(ctx, ubd)
+	}
+
+	ctx.EventManager().EmitEvent(
+		sdk.NewEvent(
+			"cancel_unbonding_delegation",
+			sdk.NewAttribute(sdk.AttributeKeyAmount, msg.Amount.String()),
+			sdk.NewAttribute(types.AttributeKeyValidator, msg.ValidatorAddress),
+			sdk.NewAttribute(types.AttributeKeyDelegator, msg.DelegatorAddress),
+			sdk.NewAttribute("creation_height", strconv.FormatInt(msg.CreationHeight, 10)),
+		),
+	)
+
+	return &types.MsgCancelUnbondingDelegationResponse{}, nil
+}
+
+// UnbondValidator defines a method for performing the status transition for
```

Comment:
Wouldn't the status transition rather be `bonded` to `unbonding`?

Reply:
I guess this is to enable a validator to jail itself as this is no longer possible with min self delegation removed. 

Indeed, the transition should be bonded to unbonding.

Reply:
that's right - added clarification [here](https://github.com/cosmos/cosmos-sdk/pull/16747/commits/750e5c4d63a2935e970637a742c39771a70f983c#diff-5ea92aa6ab2a6d865b7641169f4a530e3ec4121e427e4bc292666fec403244daR587-R590)